### PR TITLE
Rule to find `lateinit` usages

### LINKED ISSUE #106: Rule: Prohibite usage of `lateinit`

### PR DESCRIPTION
**Pull request by Mauin**
This PR introduces the `LateinitUsage` rule which will report usages of the `lateinit` modifier on properties.

Resolves #106

### CODE REVIEW

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/LateinitUsage.kt [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package io.gitlab.arturbosch.detekt.rules.style
+
+import io.gitlab.arturbosch.detekt.api.*
+import org.jetbrains.kotlin.psi.KtProperty
+import org.jetbrains.kotlin.psi.psiUtil.allChildren
+
+class LateinitUsage(config: Config = Config.empty) : Rule(config) {
+
+	override val issue = Issue(javaClass.simpleName, Severity.Style, "Usage of lateinit")
```
Comment (arturbosch) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/LateinitUsage.kt]: pls add something like "using lateinit for property initialization is error prone, try using constructor injection or delegation."

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/LateinitUsageSpec.kt [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+			class Test {
+				lateinit var first: String
+				lateinit var second: Int
+			}
+		"""
+
+		val file = compileContentForTest(code).text
+
+		it("should report lateinit usages") {
+			val rule = LateinitUsage()
```
Comment (arturbosch) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/LateinitUsageSpec.kt]: Try using LateinitUsage().lint(code)

Comment (arturbosch) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/LateinitUsageSpec.kt]: I mean you do not have to create a ktfile with compileContentForTest and then using lint on the text. Your creating two ktfiles :)

**Comment by arturbosch** [comment]
Cool new rule, thx ! :) 

I wonder if we should put this rule inside the potential bugs rule set and leave it active=false?

**Comment by vanniktech** [comment]
Sometimes you're forced to use it e.g. injecting with Dagger into activities. I feel like those cases should be whitelisted. Another one would be using ButterKnife.

**Comment by Mauin** [comment]
@vanniktech Yeah, that's why I don't want to turn this rule on by default. As for whitelisting specific cases those might differ from codebase to codebase. Maybe adding a Pattern to whitelist `lateinit` usages might make sense? (`@Inject|@BindView` in the cases you mention above)

**Comment by vanniktech** [comment]
BindView is not enough. Whitelisting per package would be cool: `butterknife.*` or `javax.inject.*`.

**Comment by arturbosch** [comment]
Hm, I'm no Android guy but aren't there any kotlin extensions for this kind of libraries where injection is done by delegation. 

For example `TornadoFx` lets me do this: `private var field by singleAssign<TextField>()` or `private val view by inject<ActionPanel>()`. 

In spring we favor constructor injection instead of field injection `class @Autowired constructor(...)`

Whitelisting annotations specified by the user would be a way to go. Excluding packages would't be too hard either (if the user does not use star imports).

**Comment by vanniktech** [comment]
There's kotterknife but it's an entire separate thing and is not as powerful as ButterKnife. For Dagger I'm not aware that there is anything like that, although I assume it can be build.

**Comment by Mauin** [comment]
I can give adding excludes a go that are set by the user. Probably along the lines what @vanniktech mentioned. 

This would add a config property:
```
LateinitUsage:
    active: false
    excludePropertiesAnnotatedWith=javax.inject.Inject|butterknife.*  
```

Does that sound sane?

**Comment by arturbosch** [comment]
Jap sounds good. 

Pls make sure to visit imports first and store matching import names in a list and check found annotations against that list instead of checking each annotation against the imports :)

**Comment by Mauin** [comment]
Pushed a new commit that works on excluding specific annotations by name. Still working on getting the Package name of an annotation. It's a bit hard to test since it seems like we're limited to compiling one class. I'd like to compile an annotation class in `package foo.bar` and then the actual test class which uses the annotation class to check if the annotation is in a specific package. Do you know of a way @arturbosch ?

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/LateinitUsage.kt [hunk 3]
```diff
@@ -1,21 +1,39 @@
...
 import org.jetbrains.kotlin.psi.KtProperty
-import org.jetbrains.kotlin.psi.psiUtil.allChildren
+import org.jetbrains.kotlin.psi.KtPsiUtil
+import org.jetbrains.kotlin.utils.ifEmpty
 
 class LateinitUsage(config: Config = Config.empty) : Rule(config) {
 
 	override val issue = Issue(javaClass.simpleName, Severity.Style, "Usage of lateinit. Using lateinit for property initialization is error prone, try using constructor injection or delegation.")
 
+	private val excludeAnnotatedProperties: List<String> = valueOrDefault(EXCLUDE_ANNOTATED_PROPERTIES, "").split("|")
```
Comment (vanniktech) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/LateinitUsage.kt]: I thought it was `,` and not `|`

Comment (Mauin) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/LateinitUsage.kt]: Yep, I read that comment only after I pushed the commit. I'll change it to `,` 👍

**Comment by arturbosch** [comment]
@Mauin checking if an annotation is in a specific package cannot be done right done as we do not resolve types and symbols before checking files. I haven't looked up how this whole things work in the compiler (it is also hardly documented, but I will get to bit later on).

So without resolving the types we just have the option to look at the imports (ignoring *-imports):
Something like
```
it("bla") {
	val code = "io.gitlab.arturbosch.detekt.api.Config"
	val pattern = "io.gitlab.arturbosch.detekt.api.*"
	val excludes = pattern.removeSuffix("*")
	val importDirective = FACTORY.createImportDirective(ImportPath.fromString(code))
	val imports = listOf(importDirective)
	val annotationsToFilter = imports.filterNot { it.isAllUnder }
			.filter { it.containsExcludedPattern(excludes) }
			.map { it.importedFqName?.shortName()?.asString() }
			.filterNotNull()

	val configAnnotation = FACTORY.createAnnotationEntry("@Config")
	val annotations = listOf(configAnnotation)
	val violations = annotations.filter { it.typeReferenceName!! in annotationsToFilter }
		
	assertThat(violations).isNotEmpty()
}


private fun KtImportDirective.containsExcludedPattern(suffix: String): Boolean {
	val importPath = importedFqName?.asString()
	val contains = importPath?.contains(suffix)
	return contains != null && contains
}
```

**Comment by Mauin** [comment]
🤔 In that case rather than parsing all imports, then all properties for its annotations while still loosing wildcard imports I think we should go with what we have now until we find a proper solution with the compiler.
This solution still allows us to filter annotations by their name and we can work on finding a way to resolve types correctly via the compiler. I think resolving types correctly might be something that will come up in the future for other rules as well. 
Once we have found a good solution we can add support for excluding annotations by package and with their full package statement?

**Comment by arturbosch** [comment]
I think it is a fair compromise, I don't know android but I don't think there will be that many annotations that writing down all of them would be hard.

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/LateinitUsage.kt [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+	private val excludeAnnotatedProperties: List<String> = valueOrDefault(EXCLUDE_ANNOTATED_PROPERTIES, "").split(",")
+
+	override fun visitProperty(property: KtProperty) {
+		val isLateinitProperty = property.modifierList?.hasModifier(KtTokens.LATEINIT_KEYWORD) ?: false
+		if (!isLateinitProperty) {
+			return
+		}
+
+		val isExcluded = property.annotationEntries
+				.map { KtPsiUtil.getShortName(it).toString() }
```
Comment (Mauin) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/LateinitUsage.kt]: Good eye on that one. Will change it now.

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/LateinitUsage.kt [hunk 5]
```diff
@@ -1,36 +1,47 @@
 package io.gitlab.arturbosch.detekt.rules.bugs
 
 import io.gitlab.arturbosch.detekt.api.*
-import org.jetbrains.kotlin.com.intellij.psi.util.PsiTreeUtil
+import jdk.nashorn.internal.ir.annotations.Ignore
```
Comment (arturbosch) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/LateinitUsage.kt]: I think this is left by mistake :)

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/LateinitUsage.kt [hunk 6]
```diff
@@ -1,36 +1,47 @@
...
 import org.jetbrains.kotlin.psi.KtPsiUtil
-import org.jetbrains.kotlin.utils.ifEmpty
 
 class LateinitUsage(config: Config = Config.empty) : Rule(config) {
 
 	override val issue = Issue(javaClass.simpleName, Severity.Style, "Usage of lateinit. Using lateinit for property initialization is error prone, try using constructor injection or delegation.")
 
 	private val excludeAnnotatedProperties: List<String> = valueOrDefault(EXCLUDE_ANNOTATED_PROPERTIES, "").split(",")
 
+	override fun visitAnnotation(annotation: KtAnnotation) {
```
Comment (arturbosch) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/LateinitUsage.kt]: This maybe too :)

**Comment by arturbosch** [comment]
:+1:

**Pull request merged**